### DISCUSSION TITLE
Database Initialization Error

### DISCUSSION THREAD

**Discussion by DejfCold** [requarks/wiki#4375] [Help / Questions] [answered]
# Environment:
- Docker  container `requarks/wiki:2`
- started by [Nomad](https://github.com/hashicorp/nomad)
- running on Rocky Linux 8.4

# The problem
Whenever the container restarts, it tries to recreate (I guess) the schema, but since the schema already exists, it fails with:
```
[MASTER] error: Database Initialization Error: create table "migrations" ("id" serial primary key, "name" varchar(255), "batch" integer, "migration_time" timestamptz) - relation "migrations" already exists
```
Then if I delete `migrations` table, it fails on `migrations_lock` and so on.

How would I go about fixing it, so I don't have to drop the tables and basically set it all up every time it restarts.
I havent's seen an issue with this nor mentioned it anywhere in the docs
Everything else seem to work fine. Except this.

My config looks like this (`db.user` and `db.pass` is resolved by Nomad and contains the actual values):
```yml
# https://github.com/Requarks/wiki/blob/dev/config.sample.yml
port: 3001
bindIp: 0.0.0.0
db:
  type: postgres
  host: 'postgresql.service.consul'
  port: 5432
  user: {{ with secret "database/creds/wiki" }}{{ .Data.username }}
  pass: {{ .Data.password }}{{ end }}
  db: wiki
  ssl: false
  sslOptions:
    auto: true
ssl:
  enabled: true
  port: 3000
  provider: custom
  format: pem
  key: /secrets/key.pem
  cert: /local/cert.pem
logLevel: info
offline: true
```

Here's the full log:
```
Loading configuration from /wiki/config.yml... OK
2021-09-01T13:28:22.597Z [MASTER] info: =======================================
2021-09-01T13:28:22.599Z [MASTER] info: = Wiki.js 2.5.201 =====================
2021-09-01T13:28:22.600Z [MASTER] info: =======================================
2021-09-01T13:28:22.600Z [MASTER] info: Initializing...
2021-09-01T13:28:23.668Z [MASTER] info: Using database driver pg for postgres [ OK ]
2021-09-01T13:28:23.674Z [MASTER] info: Connecting to database...
2021-09-01T13:28:23.837Z [MASTER] info: Database Connection Successful [ OK ]
2021-09-01T13:28:23.887Z [MASTER] error: Database Initialization Error: create table "migrations" ("id" serial primary key, "name" varchar(255), "batch" integer, "migration_time" timestamptz) - relation "migrations" already exists

```

I have a volume mounted to `/wiki/data/content`, not `/wiki/data` can this be the problem?

**Comment by DejfCold** [comment, marked as answer]
Found the problem. It was somehow caused by the `offline: true` in the config. I disabled it, removed the sideload volume restarted and it started to work.
@NGPixel, could this be documented somewhere at least? Should I create an issue?

**Comment by gregory112** [comment]
I have the same issue. I did not set `offline: true` though, but set `ha: true`, and did not mount the /data directory.

It is not mentioned in the documentation that we should mount /data directory as it is just a temporary directory (as specified in the documentation). I tried to deploy one container and it succeeds if the database was empty. When the number of containers is increased to 2, the next container will think that migration is underway and will try to recreate the tables again. So far I have not found any workaround.

**Reply by NGPixel** [reply]
As long as both containers point to the same database, additional instances won't try to run migrations. You must however start with only one container for the initial setup. Once completed, you can increase to any number of containers.

**Comment by gregory112** [comment]
Seems that the problem exists due to table owners. Just like you, I also use Vault to dynamically create database users. This becomes a problem as the first instance will create tables owned by the user for that instance. Other instances that with different users will fail with database initialization error. This happens even though I have a dedicated role that owns the wiki database and the public schema. All tables will still be created under the specific user.

Here's what I usually do with dynamic database user:
1. Create a role
2. Set the role as the owner of the database, and everything inside it
3. Let Vault (or others) create database users that is granted that role

Unfortunately, as wiki.js needs to create its own tables, by default the tables are owned by the creator. I do not know what operations that wiki.js needs to do with the tables that require owning them, but I suggest that logs should be improved to return errors regarding failed operations that require tables to be owned by a specific user.

In order to get this working, after initializing a container, I reassigned all table owners to the role I just made (with `REASSIGN OWNED`). This way other users created under that role can work. The documentation might be improved to handle this case.

**Reply by DejfCold** [reply]
Nice one! That might actually be how I fixed it and just forgot about doing that. I think so because when I look into the DB, it and all the tables are owned by the wiki role same as you describe so I have had to change it at some point.

**Reply by gregory112** [reply]
Do note that it seems wiki.js will also create a few more tables after a few settings have been applied. Accessing those tables will generate error, which according to my personal experiences, can be seen in the frontend. Changing the owner of those new tables fix the problem.
